src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs

```diff
@@ -167,11 +166,13 @@ protected ValueTask<ReadResult> StartTimingReadAsync(ValueTask<ReadResult> readA
             return readAwaitable;
         }
 
-        protected void StopTimingRead(long bytesRead)
+        protected void ReportReadBytes(long bytesRead)
         {
-            _context.TimeoutControl.BytesRead(bytesRead - _alreadyTimedBytes);
-            _alreadyTimedBytes = 0;
+            _context.TimeoutControl.BytesRead(bytesRead);
+        }
 
+        protected void StopTimingRead()
```

Comment:
I kinda liked making bytesRead a parameter to StopTimingRead(), because if you are calling StopTimingRead() you should always be calling ReportReadBytes() too. But as long as we're super careful not to miss a call to ReportReadBytes(), I'm OK with it.